cudax/include/cuda/experimental/__memory_resource/async_memory_resource.cuh

```diff
@@ -0,0 +1,312 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__MEMORY_RESOURCE_CUDA_ASYNC_MEMORY_RESOURCE_H
+#define _CUDAX__MEMORY_RESOURCE_CUDA_ASYNC_MEMORY_RESOURCE_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// cudaMallocAsync was introduced in CTK 11.2
+#if !defined(_CCCL_COMPILER_MSVC_2017) && !defined(_CCCL_CUDACC_BELOW_11_2)
+
+#  if !defined(_CCCL_CUDA_COMPILER_NVCC) && !defined(_CCCL_CUDA_COMPILER_NVHPC)
+#    include <cuda_runtime.h>
+#    include <cuda_runtime_api.h>
+#  endif // !_CCCL_CUDA_COMPILER_NVCC && !_CCCL_CUDA_COMPILER_NVHPC
+
+#  include <cuda/__memory_resource/get_property.h>
+#  include <cuda/__memory_resource/properties.h>
+#  include <cuda/__memory_resource/resource_ref.h>
+#  include <cuda/std/__concepts/__concept_macros.h>
+#  include <cuda/std/__cuda/api_wrapper.h>
+#  include <cuda/std/__new_>
+#  include <cuda/std/cstddef>
+#  include <cuda/stream_ref>
+
+#  include <cuda/experimental/__device/device_ref.cuh>
+#  include <cuda/experimental/__memory_resource/async_memory_pool.cuh>
+#  include <cuda/experimental/__stream/stream.cuh>
+
+#  if _CCCL_STD_VER >= 2014
+
+namespace cuda::experimental::mr
+{
+
+//! @brief global stream to synchronize in the synchronous interface of \c async_memory_resource
+inline ::cuda::stream_ref __async_memory_resource_sync_stream()
```

Comment:
Does it do any good to synchronize a non-default stream that the user never passes to any asynchronous CUDA operation?

```
ptr = mr.allocate(...); // syncs __async_memory_resource_sync_stream()
kernel<<<grid_size, block_size, my_stream>>>(ptr); // writes to ptr
cudaMemcpyAsync(ptr, other_mem, bytes, my_stream); // copies from ptr
mr.deallocate(ptr, ...); // syncs __async_memory_resource_sync_stream(), but ptr still being used by above
```

The above code is bad. It's the user's error. But even so, the extra synchronization that `allocate/deallocate` do has no benefit to the user's bad code that I can see.

Reply:
My opinion is that we should just drop the whole synchronous API altogether and this is only intended to work as a interim.

Isnt the code you showed buggy without the synchronization? The write on the second line would potentially access not yet valid memory if we did not synchronize beforehand, because there is no stream ordering between the asynchronous allocation inside allocate and the kernel launch

Reply:
If a synchronous `allocate` returns device memory that is not immediately available for writing on every stream, then the caller has no choice but to synchronize the device.

By definition, `allocate` must return device memory that is immediately available for writing. The code I wrote above assumes that. But my code is wrong in that it should synchronize `my_stream` before calling a synchronous `deallocate`.

Therefore, as I mentioned above, synchronizing some random stream in `allocate` and `deallocate` does nothing that I can see to help either correct or incorrect code.  

> My opinion is that we should just drop the whole synchronous API altogether and this is only intended to work as a interim.

I completely agree. But I think the only way to make this work in the interim is to call `cudaDeviceSynchronize()` or `cudaStreamSynchronize(0)`.

Reply:
Oh yeah,, but this is something we do not really want to do at all.

This is an API we want to get rid of, but still we do not want to pessimize the world for a user bug.

We cannot work around all asynchrony bugs the user might make

Reply:
So I think the redundant sync in `deallocate()` should be omitted, and the `deallocate()` docs should make clear the requirement that `ptr` is not in use on any stream before calling it.

Reply:
I dont think that is correct, because we do need the explicit sync in deallocate in case the users actually uses the memory resource synchronously.

That is a bad use case but a valid one

Reply:
Please show me an example where the sync in deallocate is required.

Reply:
I thought about suggesting the same thing, but I found one case I was not sure about:

```
ptr = async_mr.allocate(mempool_size);
// ...
async_mr.deallocate(ptr);
async_mr.allocate_async(mempool_size, some_stream);
```

Is it guaranteed that the last alloc call won't fail with OOM if there is no synchronization between `some_stream` and the internal stream used in `deallocate`?
If that works and the pool limit is updated right away when `cudaFreeAsync` is called, then I don't think there is a way to observe if deallocation is synchronous or not. We could then remove the sync from `deallocate`

Reply:
The CUDA runtime knows that `ptr` was deallocated on the *private* stream. 

By default, if the subsequent `allocate_async` fails due to fragmentation, I believe the runtime will opportunistically combine the memory freed from `ptr` with the rest of the pool and retry the allocation.  This is controlled by the CUDA MemPool attribute `cudaMemPoolReuseAllowOpportunistic`. It is enabled by default, but if it is disabled then your example would indeed fail without a synchronization.

So I guess you are right if we want the behavior to be the same whether or not the default attributes are enabled, then a synchronization is needed here.  However, we could query the attribute to decide whether or not synchronization is necessary.